druntime/src/rt/aaA.d

```diff
@@ -210,182 +215,58 @@ Bucket[] allocBuckets(size_t dim) @trusted pure nothrow
 
 private void* allocEntry(scope const Impl* aa, scope const void* pkey)
 {
-    import rt.lifetime : _d_newitemU;
     import core.stdc.string : memcpy, memset;
 
-    immutable akeysz = aa.valoff;
-    void* res = void;
-    if (aa.entryTI)
-        res = _d_newitemU(aa.entryTI);
-    else
+    immutable akeysz = aa.keysz;
+    immutable allocsz = aa.valoff + aa.valsz;
+    auto flags = (aa.flags & Impl.Flags.hasPointers) ? 0 : GC.BlkAttr.NO_SCAN;
+    if (aa.flags & Impl.Flags.hasDtors)
     {
-        auto flags = (aa.flags & Impl.Flags.hasPointers) ? 0 : GC.BlkAttr.NO_SCAN;
-        res = GC.malloc(akeysz + aa.valsz, flags);
+        assert(aa.entryTI !is null);
+        flags |= GC.BlkAttr.FINALIZE;
     }
+    auto res = GC.malloc(allocsz, flags, aa.entryTI);
 
-    memcpy(res, pkey, aa.keysz); // copy key
-    memset(res + akeysz, 0, aa.valsz); // zero value
+    memcpy(res, pkey, akeysz); // copy key
+    memset(res + akeysz, 0, allocsz - akeysz); // zero everything after
 
     return res;
 }
 
-package void entryDtor(void* p, const TypeInfo_Struct sti)
+package void finalizeEntry(void* p, TypeInfo_AssociativeArray ti)
 {
-    // key and value type info stored after the TypeInfo_Struct by tiEntry()
-    auto sizeti = __traits(classInstanceSize, TypeInfo_Struct);
-    auto extra = cast(const(TypeInfo)*)(cast(void*) sti + sizeti);
-    extra[0].destroy(p);
-    extra[1].destroy(p + talign(extra[0].tsize, extra[1].talign));
+    import rt.lifetime : unqualify;
+
+    ti.key.destroy(p);
+    if (hasDtor(unqualify(ti.value)))
+    {
+        // unfortunately, need to recalculate this every time.
+        auto valoff = talign(ti.key.tsize, ti.value.talign);
+        ti.value.destroy(p + valoff);
+    }
 }
 
 private bool hasDtor(const TypeInfo ti) pure nothrow
 {
     import rt.lifetime : unqualify;
 
     if (typeid(ti) is typeid(TypeInfo_Struct))
-        if ((cast(TypeInfo_Struct) cast(void*) ti).xdtor)
-            return true;
+        return (cast(TypeInfo_Struct) cast(void*) ti).xdtor !is null;
```

Comment:
Note: optimization. No need to test for a static array when we know it's a struct.